### PR TITLE
Get and delete agent APIs

### PR DESCRIPTION
**Pull request by mingshl**
### Description
Get and Delete Agent APIs
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/build.gradle [hunk 1]
```diff
@@ -299,6 +299,19 @@ List<String> jacocoExclusions = [
         'org.opensearch.ml.model.MLModelGroupManager',
         'org.opensearch.ml.helper.ModelAccessControlHelper',
         'org.opensearch.ml.action.models.DeleteModelTransportAction.2'
+        'org.opensearch.ml.action.models.DeleteModelTransportAction.2',
+        'org.opensearch.ml.action.tools.ListToolsTransportAction',
```
Comment (ylwu-amzn) [inline plugin/build.gradle:303]: Don't add these class to exclusion list. We should add more test to increase coverage

Comment (mingshl) [reply plugin/build.gradle]: rebase and add unit test

**Comment by mingshl** [comment]
integ failures are not related to this code changes

**Comment by ylwu-amzn** [comment]
Click [Details](https://github.com/opensearch-project/ml-commons/actions/runs/7204394295/job/19626300250?pr=1752) for the failed CI build, you can find more details
```
Execution failed for task ':opensearch-ml-plugin:jacocoTestCoverageVerification'.
> Rule violated for class org.opensearch.ml.action.agents.GetAgentTransportAction: branches covered ratio is 0.0, but expected minimum is 0.7
  Rule violated for class org.opensearch.ml.action.agents.GetAgentTransportAction: lines covered ratio is 0.1, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.action.agents.DeleteAgentTransportAction.1: lines covered ratio is 0.0, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.action.agents.DeleteAgentTransportAction: lines covered ratio is 0.3, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.rest.RestMLGetAgentAction: lines covered ratio is 0.5, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.rest.RestMLDeleteAgentAction: lines covered ratio is 0.5, but expected minimum is 0.8
```

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteActionTest.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package org.opensearch.ml.common.transport.agent;
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteActionTest.java:1]: please add license header

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package org.opensearch.ml.common.transport.agent;
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequestTest.java:1]: please add license header

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.opensearch.ml.common.transport.agent;
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java:1]: please add license header

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequestTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package org.opensearch.ml.common.transport.agent;
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequestTest.java:1]: please add license header

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponseTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
+package org.opensearch.ml.common.transport.agent;
```
Comment (dhrubo-os) [inline common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponseTest.java:1]: please add license header

File: plugin/src/test/java/org/opensearch/ml/action/agents/DeleteAgentTransportActionTests.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
+package org.opensearch.ml.action.agents;
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/action/agents/DeleteAgentTransportActionTests.java:1]: please add license header

File: plugin/src/test/java/org/opensearch/ml/action/agents/GetAgentTransportActionTests.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
+package org.opensearch.ml.action.agents;
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/action/agents/GetAgentTransportActionTests.java:1]: please add license header

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLDeleteAgentActionTests.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
+package org.opensearch.ml.rest;
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLDeleteAgentActionTests.java:1]: please add license header

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLGetAgentActionTests.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
+package org.opensearch.ml.rest;
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLGetAgentActionTests.java:1]: please add license header

Comment (mingshl) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLGetAgentActionTests.java:1]: added

**Comment by dhrubo-os** [comment]
Do you try to build in your end? fix spotless.

**Comment by mingshl** [comment]
> Do you try to build in your end? fix spotless.

Yes I did ```
./gradlew opensearch-ml-plugin:build

BUILD SUCCESSFUL in 7m 5s
45 actionable tasks: 14 executed, 31 up-to-date
```

I resolved the conversation before I made the commit 😆, Please see the new commit

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java (hunk 4 above)
Comment (dhrubo-os) [reply common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java:1]: It's showing resolved, but I don't see license header.

File: plugin/src/main/java/org/opensearch/ml/action/agents/GetAgentTransportAction.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                            new OpenSearchStatusException(
+                                "Failed to find agent with the provided agent id: " + agentId,
+                                RestStatus.NOT_FOUND
+                            )
+                        );
+                }
+            }, e -> {
+                if (e instanceof IndexNotFoundException) {
+                    log.error("Failed to get agent index", e);
+                    actionListener.onFailure(new OpenSearchStatusException("Failed to find agent", RestStatus.NOT_FOUND));
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/agents/GetAgentTransportAction.java:86]: These two doesn't quite match? What we don't get agent index or agent itself?

**Comment by ylwu-amzn** [comment]
> > Do you try to build in your end? fix spotless.
> 
> Yes I did
> 
> ```
> ./gradlew opensearch-ml-plugin:build
> 
> BUILD SUCCESSFUL in 7m 5s
> 45 actionable tasks: 14 executed, 31 up-to-date
> ```
> 
> I resolved the conversation before I made the commit 😆, Please see the new commit

You should use `./gradlew build`

**Comment by mingshl** [comment]
Paste the build failure here, this PR didn't make a change to opensearch-ml-algorithms or opensearch-ml-memory

```
* What went wrong:
A problem occurred configuring project ':opensearch-ml-algorithms'.
> Could not resolve all files for configuration ':opensearch-ml-algorithms:classpath'.
   > Could not resolve com.squareup.okio:okio-jvm:3.2.0.
     Required by:
         project :opensearch-ml-algorithms > com.diffplug.spotless:com.diffplug.spotless.gradle.plugin:6.23.0 > com.diffplug.spotless:spotless-plugin-gradle:6.23.0 > com.diffplug.spotless:spotless-lib-extra:2.43.0 > dev.equo.ide:solstice:1.7.3 > com.squareup.okhttp3:okhttp:4.11.0 > com.squareup.okio:okio:3.2.0
      > Could not resolve com.squareup.okio:okio-jvm:3.2.0.
```

```
* What went wrong:
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
Execution failed for task ':opensearch-ml-memory:test'.
```

File: common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java (hunk 4 above)
Comment (mingshl) [reply common/src/test/java/org/opensearch/ml/common/transport/agent/MLAgentGetActionTest.java:6]: added, thanks for checking!! I looks like ./gradlew spotlessApply didn't check the test packages. Added in the new commit

File: plugin/src/main/java/org/opensearch/ml/action/agents/GetAgentTransportAction.java (hunk 11 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/action/agents/GetAgentTransportAction.java]: changed to `Failed to get agent index` to log and throw the same error message. I also changed the tests too.

**Comment by dhrubo-os** [comment]
there's a merge conflict. Rebase from main.

**Comment by mingshl** [comment]
> Paste the build failure here, this PR didn't make a change to opensearch-ml-algorithms or opensearch-ml-memory
> 
> ```
> * What went wrong:
> A problem occurred configuring project ':opensearch-ml-algorithms'.
> > Could not resolve all files for configuration ':opensearch-ml-algorithms:classpath'.
>    > Could not resolve com.squareup.okio:okio-jvm:3.2.0.
>      Required by:
>          project :opensearch-ml-algorithms > com.diffplug.spotless:com.diffplug.spotless.gradle.plugin:6.23.0 > com.diffplug.spotless:spotless-plugin-gradle:6.23.0 > com.diffplug.spotless:spotless-lib-extra:2.43.0 > dev.equo.ide:solstice:1.7.3 > com.squareup.okhttp3:okhttp:4.11.0 > com.squareup.okio:okio:3.2.0
>       > Could not resolve com.squareup.okio:okio-jvm:3.2.0.
> ```
> 
> ```
> * What went wrong:
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
> Execution failed for task ':opensearch-ml-memory:test'.
> ```

reruning Linux 17 and flaky tests are resolved. [Success Build](https://github.com/opensearch-project/ml-commons/actions/runs/7227840058/job/19697150831?pr=1752)

**Comment by ylwu-amzn** [comment]
```
> Task :opensearch-ml-plugin:spotlessJava FAILED
Step 'removeUnusedImports' found problem in 'src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java':
341:101: error: ')' expected
```

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+    public static MLAgentDeleteRequest fromActionRequest(ActionRequest actionRequest) {
+        if (actionRequest instanceof MLAgentDeleteRequest) {
+            return (MLAgentDeleteRequest)actionRequest;
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             OutputStreamStreamOutput osso = new OutputStreamStreamOutput(baos)) {
+            actionRequest.writeTo(osso);
+            try (StreamInput input = new InputStreamStreamInput(new ByteArrayInputStream(baos.toByteArray()))) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequest.java:63]: let's add test for this section

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    public static MLAgentGetRequest fromActionRequest(ActionRequest actionRequest) {
+        if (actionRequest instanceof MLAgentGetRequest) {
+            return (MLAgentGetRequest) actionRequest;
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             OutputStreamStreamOutput osso = new OutputStreamStreamOutput(baos)) {
+            actionRequest.writeTo(osso);
+            try (StreamInput input = new InputStreamStreamInput(new ByteArrayInputStream(baos.toByteArray()))) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequest.java:64]: let's add test for this section

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponse.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    public static MLAgentGetResponse fromActionResponse(ActionResponse actionResponse) {
+        if (actionResponse instanceof MLAgentGetResponse) {
+            return (MLAgentGetResponse) actionResponse;
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             OutputStreamStreamOutput osso = new OutputStreamStreamOutput(baos)) {
+            actionResponse.writeTo(osso);
+            try (StreamInput input = new InputStreamStreamInput(new ByteArrayInputStream(baos.toByteArray()))) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponse.java:54]: let's add test for this section

**Review by ylwu-amzn — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
Don't have major concern. Code coverage is not enough, please add more unit test.
Screenshot 2023-12-16 at 11 21 27 AM

**Review by dhrubo-os — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponse.java (hunk 14 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetResponse.java:54]: Added in #1791

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequest.java (hunk 13 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentGetRequest.java:64]: Added in #1791

File: common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequest.java (hunk 12 above)
Comment (mingshl) [reply common/src/main/java/org/opensearch/ml/common/transport/agent/MLAgentDeleteRequest.java:63]: Added in #1791

**Comment by mingshl** [comment]
> Don't have major concern. Code coverage is not enough, please add more unit test. Screenshot 2023-12-16 at 11 21 27 AM

Added in #1791

**Pull request merged**